rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/topic_cache.hpp

```diff
@@ -21,49 +21,79 @@
 #include <set>
 #include <sstream>
 #include <string>
+#include <tuple>
 #include <unordered_map>
 #include <utility>
 #include <vector>
 
 #include "fastrtps/participant/Participant.h"
 #include "fastrtps/rtps/common/Guid.h"
 #include "fastrtps/rtps/common/InstanceHandle.h"
+#include "qos.hpp"
 #include "rcpputils/thread_safety_annotations.hpp"
 #include "rcutils/logging_macros.h"
 
 typedef eprosima::fastrtps::rtps::GUID_t GUID_t;
 
+/**
+ * A data structure that encapsulates all the data associated with a publisher
+ * or subscription by the topic it publishers or subscribers to
+ */
+struct TopicData
+{
+  GUID_t participant_guid;
+  GUID_t entity_guid;
+  std::string topic_type;
+  rmw_qos_profile_t qos_profile;
+};
+
 /**
  * Topic cache data structure. Manages relationships between participants and topics.
  */
 class TopicCache
 {
 private:
-  typedef std::map<GUID_t,
-      std::unordered_map<std::string, std::vector<std::string>>> ParticipantTopicMap;
-  typedef std::unordered_map<std::string, std::vector<std::string>> TopicToTypes;
+  using TopicToTypes = std::unordered_map<std::string, std::vector<std::string>>;
+  using ParticipantTopicMap = std::map<GUID_t, TopicToTypes>;
+  using TopicNameToTopicData = std::unordered_map<std::string, std::vector<TopicData>>;
 
   /**
-   * Map of topic names to a vector of types that topic may use.
+   * Map of topic names to TopicData. Where topic data is vector of tuples containing
+   * participant GUID, topic type and the qos policy of the respective participant.
    * Topics here are represented as one to many, DDS XTypes 1.2
-   * specifies application code 'generally' uses a 1-1 relationship.
+   * specifies application code that 'generally' uses a 1-1 relationship.
    * However, generic services such as logger and monitor, can discover
    * multiple types on the same topic.
-   *
    */
-  TopicToTypes topic_to_types_;
+  TopicNameToTopicData topic_name_to_topic_data_;
 
   /**
    * Map of participant GUIDS to a set of topic-type.
    */
   ParticipantTopicMap participant_to_topics_;
 
   /**
-   * Helper function to initialize a topic vector.
+   * Helper function to initialize an empty TopicData for a topic name.
    *
-   * @param topic_name
+   * @param topic_name the topic name for which the TopicNameToTopicData map should be initialised.
+   * @param topic_name_to_topic_data the map to initialise.
```

Comment:
@jaisontj nit: here and below, both `initialize` and `initialise` are valid, but let's standardize to one form

Reply:
Standardized on "initialize".